src/lib/input/input-container.ts

```diff
@@ -124,7 +124,21 @@ export class MdInputDirective implements AfterContentInit {
    */
   @Output() _placeholderChange = new EventEmitter<string>();
 
-  get empty() { return (this.value == null || this.value === '') && !this._isNeverEmpty(); }
+  get empty() {
+    if (this._isNeverEmpty()) {
+      return false;
+    }
+    if (this.value == null || this.value === '') {
+      // Check if this is an <input> element that contains bad input.
+      // If so, we know that it only appears empty because the value failed to parse.
+      if (this._elementRef.nativeElement instanceof HTMLInputElement &&
+          this._elementRef.nativeElement.validity.badInput) {
```

Comment:
I tried, but you can't programaticaly set the value to a bad input. I'd have to simulate key presses which sounded like more trouble than its worth